File: day009_string_advanced/related_valid_palindrome/solution.java
import java.util.Scanner;

class Solution {

    // O(N) Time, O(1) Space - Optimal (Two Pointers with in-place filtering)
    public static void solve_valid_palindrome() {
        Scanner scanner = new Scanner(System.in);
        if (!scanner.hasNextLine()) return;
        String S = scanner.nextLine();
        
        int left = 0;
        int right = S.length() - 1;

        while (left < right) {
            // Skip non-alphanumeric characters from the left
            while (left < right && !Character.isLetterOrDigit(S.charAt(left))) {
                left++;
            }
            
            // Skip non-alphanumeric characters from the right
            while (left < right && !Character.isLetterOrDigit(S.charAt(right))) {
                right--;
            }

            // Compare characters (case-insensitive)
            if (left < right && Character.toLowerCase(S.charAt(left)) != Character.toLowerCase(S.charAt(right))) {
                System.out.println("NO");
                scanner.close();
                return;
            }

            // Move pointers inward
            left++;
            right--;
        }
        
        System.out.println("YES");
        scanner.close();
    }

    // O(N) Time, O(N) Space - Brute Force (Preprocessing)
    public static void solve_valid_palindrome_brute() {
        Scanner scanner = new Scanner(System.in);
        if (!scanner.hasNextLine()) return;
        String S = scanner.nextLine();

        // 1. Filter and convert (O(N) time and space)
        StringBuilder filtered = new StringBuilder();
        for (char c : S.toCharArray()) {
            if (Character.isLetterOrDigit(c)) {
                filtered.append(Character.toLowerCase(c));
            }
        }
        
        String filteredString = filtered.toString();
        
        // 2. Check if string equals its reverse (O(N) time)
        String reversedString = filtered.reverse().toString();

        if (filteredString.equals(reversedString)) {
            System.out.println("YES");
        } else {
            System.out.println("NO");
        }
        scanner.close();
    }

    public static void main(String[] args) {
        solve_valid_palindrome();
    }
}

File: day009_string_advanced/related_valid_palindrome/solution.py
import sys

def solve_valid_palindrome():
    try:
        S = sys.stdin.readline().strip()
    except:
        return

    # Optimized: O(N) Time, O(1) Space - Two Pointers with in-place filtering
    
    left = 0
    right = len(S) - 1
    
    while left < right:
        # Move left pointer past non-alphanumeric characters
        while left < right and not S[left].isalnum():
            left += 1
            
        # Move right pointer past non-alphanumeric characters
        while left < right and not S[right].isalnum():
            right -= 1
            
        # Compare characters (case-insensitive)
        if S[left].lower() != S[right].lower():
            print("NO")
            return
            
        # Move both pointers inward
        left += 1
        right -= 1
        
    print("YES")

# Brute Force: O(N) Time, O(N) Space - Preprocessing
def solve_valid_palindrome_brute():
    try:
        S = sys.stdin.readline().strip()
    except:
        return
    
    # Preprocess: Filter and convert to lower case (O(N) time and O(N) space)
    filtered_s = "".join(char.lower() for char in S if char.isalnum())
    
    # Check if string equals its reverse (O(N) time)
    if filtered_s == filtered_s[::-1]:
        print("YES")
    else:
        print("NO")

if __name__ == "__main__":
    solve_valid_palindrome()